Add proof verification example in Solidity to README (#6)
OpenZeppelin/merkle-tree: README.md

In practice this might be done in a frontend application prior to submitting the proof on-chain, with the address looked up being that of the connected wallet.

### Validating a Proof in Solidity

> Once you've gotten the proof, it can be validated in Solidity using [`MerkleProof`] with the following example:

Reviewer: ```suggestion
Once the proof has been generated, it can be validated in Solidity using [`MerkleProof`] as in the following example:
```

```solidity
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";

contract Verifier {
    bytes32 private root;

    constructor(bytes32 _root) {
        // (1)
        root = _root;
    }

    function verify(
        bytes32[] memory proof,
        address addr,
        uint256 amount

>     ) public view returns (bool) {
>         // (2)
>         bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(addr, amount))));
>         // (3)
>         return MerkleProof.verify(proof, root, leaf);

Reviewer: Normally the proof will be verified as a prerequisite to doing some action, so let's add a fourth item where we say there would be some other thing happening (not sure how best to express that).

There's something we should also mention, though we don't have a great recommendation yet, but usually you want to mark a leaf as "used" so it can't be reused. We can just put a note in parentheses saying to consider that there may need be a mechanism to prevent reuse of a leaf/proof.

```suggestion
    ) public {
        // (2)
        bytes32 leaf = keccak256(bytes.concat(keccak256(abi.encode(addr, amount))));
        // (3)
        require(MerkleProof.verify(proof, root, leaf), "Invalid proof");
        // (4)
        // ...
```

Author: > let's add a fourth item where we say there would be some other thing happening
Done, however, I decided to not use a `require` and instead express the verification as a `bool verified` value. I think it fits better in expressing that the value might be used for further operations. Let me know if that works.

> We can just put a note in parentheses saying to consider that there may need be a mechanism to prevent reuse of a leaf/proof.
Added!

    }
}
```

> 1. Load the tree's root to your contract.

Reviewer: ```suggestion
1. Store the tree root in your contract.
```

> 2. Compute the leaf's value for the provided `addr` and `amount` ABI encoded values.

Reviewer: Let's add a new heading in the "Standard Merkle Trees" section, `### Leaf Hash`, just above "From the last three ...". We can link it here:

```suggestion
2. Compute the [leaf hash](#leaf-hash) for the provided `addr` and `amount` ABI encoded values.
```

3. Verify it using [`MerkleProof`]'s `verify` function.

> **Note**

> > The provided example is not suitable for production purposes since it assumes the `root` will never be changed. Consider adding a mechanism to securely update the `root`

Reviewer: This isn't true! Sometimes you can use an immutable root, and people should favor that if possible to keep things permissionless. I would just remove this note for now to keep it simpler.

Author: Gotcha! I'm removing it
Do we have an opinion on these sorts of notes? I'd preferred to add it instead of assuming it'd be safe as an example

Reviewer: The spirit of this note I agree with, but the content I disagree with. If something is "not suitable for production purposes", it makes sense to make a note of it, but I don't think it's the case here. The code seems perfectly suitable for many production use cases.

Author: Amazing, now I got it. I'll continue trying to add notes were it may be needed 😄
